Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,135 @@
+module Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+  ( SBVector(..)
+  , SBMatrix(..)
+  , mult
+  , size
+  , linearSolve
+  , testLinearSolver
+  ) where
+
+import qualified Data.List as L
+import qualified Data.IntSet as S
+import qualified Data.Vector as V
+import qualified Data.Vector.Unboxed as U
+import Debug.Trace
+import Data.Semigroup()
+import System.Random
+import System.IO.Unsafe
+import GHC.Clock
+import Data.Bit
+import Data.Bits
+
+-- Sparse Binary Vector
+newtype SBVector = SBVector
+  { set :: S.IntSet
+  } deriving (Show, Eq)
+
+-- Vectors form a group (in particular a monoid) under addition.
+instance Semigroup SBVector where
+  SBVector v1 <> SBVector v2 = SBVector ((v1 S.\\ v2) <> (v2 S.\\ v1))
+
+instance Monoid SBVector where
+  mempty = SBVector mempty
+
+dot :: SBVector -> SBVector -> Bit
+dot (SBVector v1) (SBVector v2) = Bit (even (S.size (v1 `S.intersection` v2)))
+-- Sparse Binary Matrix
+newtype SBMatrix = SBMatrix (V.Vector SBVector) deriving (Show)
+
+mult :: SBMatrix -> SBVector -> SBVector
+mult (SBMatrix matrix) (SBVector vector) = foldMap (matrix V.!) (S.toList vector)
```

Comment:
`foldMap` will allocate a lot of intermediate sets in memory. Another approach is to flip bits in a bit vector and then convert it back to set:

```haskell
mult :: SBMatrix -> SBVector -> SBVector
mult (SBMatrix matrix) (SBVector vector) = runST $ do
  -- it would be better to cache maximal length somewhere
  let len = getMax (fromJust (foldMap (fmap (Max . fst) . S.maxView . set) matrix)) + 1
  vs <- MU.new len
  traverse_ (traverse_ (unsafeFlipBit vs) . S.toList . set . (matrix V.!)) (S.toList vector)
  ws <- U.unsafeFreeze vs
  pure $ SBVector $ S.fromDistinctAscList $ listBits ws
```

Reply:
Would that be very similar to implement dense matrices?

Reply:
See my comment below: https://github.com/Bodigrim/arithmoi/pull/208#discussion_r449803481

Reply:
`Max` / `getMax` are from `Data.Semigroup`.
`traverse_` is from `Data.Foldable`.